src/storage/src/source/persist_source.rs

```diff
@@ -339,29 +215,77 @@ where
                     data_shard.clone(),
                 )
                 .await
-                .expect("could not open persist shard");
+                .expect("could not open persist shard")
+                .batch_fetcher();
 
             initial_capabilities.clear();
 
-            let mut output_handle = output.activate();
+            let mut output_handle = update_output.activate();
+            let mut seqno_lease_output_handle = seqno_lease_output.activate();
 
-            while let Some((cap, data)) = input.next() {
-                let cap = cap.retain();
+            while let Some((cap, data)) = fetcher_input.next() {
+                let update_cap = cap.delayed_for_output(cap.time(), update_output_port);
+                let mut update_session = output_handle.session(&update_cap);
+
+                let seqno_cap = cap.delayed_for_output(cap.time(), seqno_lease_port);
+                let mut seqno_session = seqno_lease_output_handle.session(&seqno_cap);
                 for (_idx, batch) in data.iter() {
-                    let mut updates = read
+                    let mut updates = fetcher
                         .fetch_batch(batch.clone())
                         .await
                         .expect("shard_id generated for sources must match across all workers");
 
-                    let mut session = output_handle.session(&cap);
-                    session.give_vec(&mut updates);
+                    update_session.give_vec(&mut updates);
+                    seqno_session.give(batch.seqno());
                 }
             }
             false
         }),
     );
 
-    let (ok_stream, err_stream) = output_stream.ok_err(|x| match x {
+    // This operator is meant to only run on the chosen worker. All workers will
+    // exchange their fetched batches' `SeqNo` back to the leasor.
+    let mut seqno_leasor_builder = OperatorBuilder::new(
+        format!("persist_source {:?}: SeqNo leasor", source_id),
+        scope.clone(),
+    );
+
+    // Exchange all `SeqNo`s back to the chosen worker/leasor.
+    let seqno_leasor_dist = move |&_: &SeqNo| u64::cast_from(chosen_worker);
+    let mut seqno_leasor_input = seqno_leasor_builder
+        .new_input(&seqno_lease_output_stream, Exchange::new(seqno_leasor_dist));
+
+    seqno_leasor_builder.build_async(
+        scope.clone(),
+        async_op!(|initial_capabilities, _frontiers| {
+            initial_capabilities.clear();
+
+            // The chosen worker is the leasor.
+            if worker_index != chosen_worker {
+                trace!("We are not the SeqNo leasor of {:?}, exiting...", source_id);
+                return false;
+            }
+
+            let subscription = Arc::clone(&subscribe);
+
+            // Ensure subscription has been generated.
+            while subscription.lock().await.is_none() {
```

Comment:
could use a oneshot channel instead of an Option inside the Arc<Mutex<>> to avoid the sleep loop

Reply:
Using the channel obviates the need to do anything with the mutex 🎉